opal/mca/base/mca_base_source.c

```diff
@@ -0,0 +1,242 @@
+ * Copyright (c) 2017      IBM Corporation. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "mca_base_source.h"
+#include "mca_base_vari.h"
+
+#include <stddef.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+
+#include "opal/class/opal_pointer_array.h"
+#include "opal/class/opal_hash_table.h"
+#include "opal/util/clock_gettime.h"
+
+static opal_pointer_array_t registered_sources;
+static bool mca_base_source_initialized = false;
+static int source_count;
+
+static opal_mutex_t mca_base_source_lock = OPAL_MUTEX_STATIC_INIT;
+
+int mca_base_source_default_source = -1;
+
+static uint64_t mca_base_source_default_time_source (void)
+{
+    uint64_t time_value;
+    struct timespec current;
+
+    (void)opal_clock_gettime(&current);
+    time_value = 1000000000ul * current.tv_sec + current.tv_nsec;
+
+    return time_value;
+}
+
+static uint64_t mca_base_source_default_time_source_ticks (void)
+{
+    struct timespec spec;
+    if (0 == opal_clock_getres(&spec)){
+        return (uint64_t)(spec.tv_sec + spec.tv_nsec);
+    } else {
+        /* guess */
+        return 1000000;
+    }
+}
+
+/***************************************************************************************************/
+
+int mca_base_source_init (void)
+{
+    int ret = OPAL_SUCCESS;
+
+    OPAL_THREAD_LOCK(&mca_base_source_lock);
+    if (false == mca_base_source_initialized) {
+        mca_base_source_initialized = true;
+
+        OBJ_CONSTRUCT(&registered_sources, opal_pointer_array_t);
+        opal_pointer_array_init(&registered_sources, 16, 512, 16);
+
+        mca_base_source_default_source = mca_base_source_register ("opal", "mca", "base", "default_source",
+                                                                   "Default source for MCA events", true,
+                                                                   mca_base_source_default_time_source,
+                                                                   mca_base_source_default_time_source_ticks ());
+
+    }
+    OPAL_THREAD_UNLOCK(&mca_base_source_lock);
+
+    return ret;
+}
+
+int mca_base_source_finalize (void)
+{
+    int i;
+
+    OPAL_THREAD_LOCK(&mca_base_source_lock);
+    if (true == mca_base_source_initialized)  {
+        for (i = 0 ; i < source_count ; ++i) {
+            mca_base_source_t *source = opal_pointer_array_get_item (&registered_sources, i);
+            if (source) {
+                OBJ_RELEASE(source);
+            }
+        }
+
+        source_count = 0;
+
+        OBJ_DESTRUCT(&registered_sources);
+        mca_base_source_initialized = false;
+    }
+    OPAL_THREAD_UNLOCK(&mca_base_source_lock);
+
+    return OPAL_SUCCESS;
+}
+
+/***************************************************************************************************/
+
+mca_base_source_t *mca_base_source_get (int source_index)
+{
+    return opal_pointer_array_get_item (&registered_sources, source_index);
```

Comment:
In some places in this file, accessing `registered_sources` is protected by `mca_base_source_lock`, but other places -- like here -- it's not.  Is there a reason for the discrepancy?

Reply:
`mca_base_source_lock` is used during source init and finalize, but the source data should not change after initialization.

Reply:
Addressed with https://github.com/open-mpi/ompi/commit/c9b35e2ac9a780d16a05bfc9b82b4aef5ab0a970